### PR TITLE
canbus: Convert canbus driver and subsys to new timeout API

### PR DESCRIPTION
**Pull request by alexanderwachter**
Convert all canbus related API/samples/tests/subsys
to the new timeout API with k_timeout_t.

### CODE REVIEW

**Review by carlescufi — APPROVED** [review]
Good to see use of `k_timeout_t` in public APIs. I have opted for the same approach in the modem command handler, but most other APIs have instead kept `s32_t`.

**Review by tbursztyka — APPROVED** [review]

**Review by jukkar — APPROVED** [review]

File: subsys/net/l2/canbus/6locan.c [hunk 1]
```diff
@@ -145,22 +145,17 @@ static void canbus_st_min_timeout(struct _timeout *t)
...
 {
-	s32_t time_ms;
-
 	/* According to ISO 15765-2 stmin should be 127ms if value is corrupt */
 	if (stmin > NET_CAN_STMIN_MAX ||
 	    (stmin > NET_CAN_STMIN_MS_MAX && stmin < NET_CAN_STMIN_US_BEGIN)) {
-		time_ms = K_MSEC(NET_CAN_STMIN_MS_MAX);
+		return K_MSEC(NET_CAN_STMIN_MS_MAX);
 	} else if (stmin >= NET_CAN_STMIN_US_BEGIN) {
-		/* This should be 100us-900us but zephyr can't handle that */
```
Comment (andyross) [inline subsys/net/l2/canbus/6locan.c:157]: But now we can!  (Though 100us is in fact the default tick period on most non-nRF hardware platforms now, so this is still pushing the limits of what you can expect to work).

**Review by andyross — APPROVED** [review]

File: drivers/can/can_stm32.c [hunk 2]
```diff
@@ -526,10 +525,10 @@ int can_stm32_recover(struct device *dev, s32_t timeout)
 
 	can_leave_init_mode(can);
 
-	start_time = k_cycle_get_32();
+	start_time = k_uptime_ticks();
 
 	while (can->ESR & CAN_ESR_BOFF) {
-		if (k_cycle_get_32() - start_time >= CAN_BOFF_RECOVER_TIMEOUT) {
+		if (k_uptime_ticks() - start_time >= timeout.ticks) {
```
Comment (karstenkoenig) [inline drivers/can/can_stm32.c:531]: Does this not need to check for the special value K_FOREVER?

File: samples/net/sockets/can/src/main.c [hunk 3]
```diff
@@ -60,7 +60,7 @@ static void tx(int *can_fd)
 	LOG_DBG("Sending CAN data...");
 
 	while (1) {
-		ret = send(fd, &frame, sizeof(frame), K_FOREVER);
+		ret = send(fd, &frame, sizeof(frame), SYS_FOREVER_MS);
```
Comment (karstenkoenig) [inline samples/net/sockets/can/src/main.c:63]: Will this not pass SYS_FOREVER_MS as the flags for the socket send() function?

File: samples/drivers/can/src/main.c [hunk 4]
```diff
@@ -257,7 +257,7 @@ void main(void)
 	while (1) {
 		change_led_frame.data[0] = toggle++ & 0x01 ? SET_LED : RESET_LED;
 		/* This sending call is none blocking. */
-		can_send(can_dev, &change_led_frame, SYS_FOREVER_MS,
+		can_send(can_dev, &change_led_frame, K_FOREVER,
```
Comment (karstenkoenig) [inline samples/drivers/can/src/main.c:260]: In can.h it is still documented to use `SYS_FOREVER_MS` for forever

File: drivers/can/can_stm32.c (hunk 2 above)
Comment (alexanderwachter) [reply drivers/can/can_stm32.c:531]: You are right. Good catch!

File: subsys/net/l2/canbus/6locan.c (hunk 1 above)
Comment (alexanderwachter) [reply subsys/net/l2/canbus/6locan.c:157]: That's right!

File: samples/net/sockets/can/src/main.c (hunk 3 above)
Comment (alexanderwachter) [reply samples/net/sockets/can/src/main.c:63]: Yes. @jukkar what does the socket expect here?

File: samples/drivers/can/src/main.c (hunk 4 above)
Comment (alexanderwachter) [reply samples/drivers/can/src/main.c:260]: Changed documentation to use K_FOREVER

File: samples/net/sockets/can/src/main.c (hunk 3 above)
Comment (jukkar) [reply samples/net/sockets/can/src/main.c:63]: The SYS_FOREVER_MS which has a value -1, is currently ignored as the value is expected to be a flag, and the -1 is not a proper flag value. The default in send() is to wait forever so the end result is the same. If one does not want to wait at all, then MSG_DONTWAIT value should be supplied here or socket can be made non blocking.
Best here, would be to set the flags to value 0 in order not to confuse the reader.

Comment (alexanderwachter) [reply samples/net/sockets/can/src/main.c]: Done

**Review by karstenkoenig — APPROVED** [review]
The interfaces look much better with the proper timeouts, thanks!

**Pull request merged**